src/System.Management.Automation/engine/debugger/debugger.cs

```diff
@@ -2500,6 +2518,39 @@ public override void SetDebuggerStepMode(bool enabled)
             }
         }
 
+        /// <summary>
+        /// Breaks into the debugger.
+        /// </summary>
+        /// <param name="triggerObject">The object that triggered the breakpoint, if there is one.</param>
+        internal override void Break(object triggerObject = null)
+        {
+            if (!IsDebugHandlerSubscribed &&
+                (UnhandledBreakpointMode == UnhandledBreakpointProcessingMode.Ignore))
+            {
+                // No debugger attached and runspace debugging is not enabled.  Enable runspace debugging here
+                // so that this command is effective.
+                UnhandledBreakpointMode = UnhandledBreakpointProcessingMode.Wait;
+            }
+
+            // Store the triggerObject so that we can add it to PSDebugContext
+            TriggerObject = triggerObject;
+
+            // Set debugger to step mode so that a break can occur.
+            SetDebuggerStepMode(true);
```

Comment:
`-ErrorAction break` will set the debugger mode to be `Enabled`, and after that the debugger mode will stay `Enabled`. That means once a `ActionPreference.Break` is hit, the session will never be able to run script in optimized mode anymore.
This feels like a concern to me. I can imagine one scenario where people uses `-ErrorAction Break` -- I don't expect an error will be thrown but want to debug in case the unexpected happens, but after debugging (quit the debugger prompt), I would like the rest of my scripts continue to run as efficient as it can (in optimized mode). That scenario will be compromised as the debugger mode will always be `Enabled` after the `-ErrorAction Break` is hit, and thus the `Compiler` will generated unoptimized code for the rest time of the session.

`Set-PSBreakpoint` turns the debugger mode to `Enabled`, and `Remove-PSBreakpoint` turns it back to `Disabled` after removing the last breakpoint. Maybe what we need is to reset the debugger mode after quitting the debugger prompt if we broke in the debugger because of `ActionPreference.Break`. Just an idea, totally thinking aloud here.

Reply:
Interesting idea.

A few relevant thoughts:

1. `Remove-PSBreakpoint` only turns the debugger mode back to disabled if it is invoked while you are not debugging (this was identified in #9824 and fixed as part of #9825) -- as a command-line debugger, I routinely add breakpoints, debug things, remove breakpoints while in the debugger, and then step through things. This is the right thing to do, because otherwise I can remove a breakpoint, then step, but the debugger doesn't step, it continues instead. Based on what you mention here, it sounds like this would also leave my debugger mode as `Enabled`.

1. A PR that I plan to submit once this is merged in is to add `-Condition <ScriptBlock>` as a parameter to `Wait-Debugger` (while also updating it to support pipeline input), so that you can enter the debugger conditionally with that command. I would expect the condition to be evaluated and that event to trigger even if the debug mode is `Disabled`.

1. With `ActionPreference.Break`, you can have dynamic breakpointing for all messages, so if I detach from the debugger after I processed one dynamic message trigger, I would want other messages to trigger breakpoints as well.

1. Issue #9336 highlights how detaching from a job prevents dynamic breakpoints like `Wait-Debugger` from triggering. Those should always trigger, so we need to be careful about what we do when we detach a debugger. I'm curious what the cause of #9336 is: what are we doing when we detach that prevents `Wait-Debugger` from triggering in the first place -- does the compiler enable debugger mode before a script with `Wait-Debugger` is even run? Or are we doing more than detach when we detach from a job? I just haven't dug into it yet to figure that out, but I think that's an important detail to know to make sure that we persist debug mode when appropriate.

That said, we should run optimized when possible, but rather than relying on `Remove-PSBreakpoint`, where would be an appropriate place to safely disable debugger mode? Can we do it when we're back at the top-level prompt if there are no explicit breakpoints, since nothing is running at that point?

Reply:
We can experiment with this, but it seems like the right place to do this is when processing a debugger 'continue' or 'detach' command.  If no breakpoints are set then the debugger can be disabled until Wait-Debugger or ErrorAction break occurs.

Reply:
@PaulHigin As a follow-up PR though once this one is merged, right? Or would you want some investigation done as part of this PR?

Reply:
I'm fine with addressing this in a follow-up PR. Opened issue https://github.com/PowerShell/PowerShell/issues/10380 to track this.